src/protocol.h

```diff
@@ -294,6 +296,7 @@ enum ServiceFlags : uint64_t {
  * Thus, generally, avoid calling with peerServices == NODE_NONE.
  */
 static ServiceFlags GetDesirableServiceFlags(ServiceFlags services) {
+    if (services & NODE_NETWORK_LIMITED && initialBlockDownloadCompleted) return ServiceFlags(NODE_NETWORK_LIMITED | NODE_WITNESS);
```

Comment:
On https://github.com/bitcoin/bitcoin/pull/10387/files#diff-eff7adeaec73a769788bb78858815c91R1625 you are defining a limitedPeer as one which does not have `NODE_NETWORK` AND which have `NODE_NETWORK_LIMITED`. Which is different from this logic.

Reply:
I can't follow you comment.
`pfrom->fLimitedNode` is set to true if the peer does not signal `NODE_NETWORK` but signals `NODE_NETWORK_LIMITED`.
At this point `GetDesirableServiceFlags()`, we should return `NODE_NETWORK_LIMITED` (not `NODE_NETWORK`) if we are out of IBD. I think this is correct,...

Reply:
In other words, this is what I would have done:

```
var limited =  (!(services & NODE_NETWORK) && (services & NODE_NETWORK_LIMITED));
if (limited && initialBlockDownloadCompleted) return ServiceFlags(NODE_NETWORK_LIMITED | NODE_WITNESS);
return ServiceFlags(NODE_NETWORK | NODE_WITNESS);
```

The reason is that this is how you define if a peer is limited from his services flags [before]( https://github.com/bitcoin/bitcoin/pull/10387/files#diff-eff7adeaec73a769788bb78858815c91R1625), I don't understand why you are using different logic here.

Reply:
I don't think we should check for `!(services & NODE_NETWORK)` because we also want to connect to NODE_NETWORK peers... but we are also happy by connecting to a peer that has `NODE_NETWORK_LIMITED` but not `NODE_NETWORK`.

Reply:
Imagine the following scenario:

* You are not in IBD, connect to peers only with `NODE_NETWORK|NODE_NETWORK_LIMITED`
* For some reason, you start being in IBD once again.

Now your node is stuck because of [this](https://github.com/bitcoin/bitcoin/pull/10387/files#diff-eff7adeaec73a769788bb78858815c91R3598).

An alternative fix would be to only check for `NODE_NETWORK_LIMITED` [here](https://github.com/bitcoin/bitcoin/pull/10387/files#diff-eff7adeaec73a769788bb78858815c91R1625).

Reply:
> In general I think it is weird that we treat differently NODE_NETWORK|NODE_NETWORK_LIMITED peer from NODE_NETWORK peers.

I think we have to. Otherwise we would never connect to pruned peers... what we want is also considering connecting to pruned if we are not deeper then 144 blocks from the expected tip.

> For some reason, you start being in IBD once again.

I think there you are right. When switching back to IBD below say 144*1.5 blocks from the tip, we eventually hit a peer stalling us. Though it should be detected by the stalling logic and should get disconnected. The free slow should be then be re-used with a peer that must have `NODE_NETWORK` (because `GetDesirableServiceFlags()` will required `NODE_NETWORK` again at this point).

But I agree that disconnecting pruned (`NODE_NETWORK_LIMITED` only) peers once we fallback to IBD should be something we implement.

Reply:
`NODE_NETWORK|NODE_NETWORK_LIMITED` makes no sense to me. I was under the impression that if a peer was serving only limited block it would be `NODE_NETWORK_LIMITED`. If the peer can serve more, it would be `NODE_NETWORK`.

So what is the meaning of a peer `NODE_NETWORK|NODE_NETWORK_LIMITED` ? I would argue that this mean they can serve historical blocks, so they should be considered as if they were `NODE_NETWORK` only.

Reply:
As of BIP159, `NODE_NETWORK_LIMITED` signals that a peer has the last 288 blocks. This is "always" true for a NODE_NETWORK peer. This was chosen because services are a bit-set and bits should not depend on each other.

Conclusion:
* `NODE_NETWORK` (unpruned peer, can serve the complete history)
* `NODE_NETWORK_LIMITED ^NODE_NETWORK` (no NODE_NETWORK) (can serve the last 288 blocks but can't serve the complete history)
* `NODE_NETWORK | NODE_NETWORK_LIMITED` (can serve the complete history and the last 288 blocks [which is obvious]

Reply:
OK I got it now. Thanks for explanation, this is fine to me.